File: backend/app/security/encryption.py
import base64
import json
import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

_key_cache = None

def get_key() -> bytes:
    global _key_cache
    if _key_cache:
        return _key_cache

    key_str = os.getenv("ENCRYPTION_KEY", "")
    if not key_str:
        from app.config import get_settings
        settings = get_settings()
        key_str = settings.ENCRYPTION_KEY

    if not key_str:
        raise ValueError("ENCRYPTION_KEY not set in environment or .env")

    _key_cache = base64.b64decode(key_str)
    return _key_cache


def encrypt(data: dict) -> str:
    key = get_key()
    aesgcm = AESGCM(key)
    nonce = os.urandom(12)
    plaintext = json.dumps(data).encode()
    ciphertext = aesgcm.encrypt(nonce, plaintext, None)
    return base64.b64encode(nonce + ciphertext).decode()


def decrypt(token: str) -> dict:
    key = get_key()
    aesgcm = AESGCM(key)
    raw = base64.b64decode(token)
    nonce = raw[:12]
    ciphertext = raw[12:]
    plaintext = aesgcm.decrypt(nonce, ciphertext, None)
    return json.loads(plaintext)

File: frontend/src/utils/encryption.js
const ALGORITHM = { name: 'AES-GCM', length: 256 }

function base64ToArrayBuffer(base64) {
  const binary = atob(base64)
  const bytes = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i)
  }
  return bytes.buffer
}

function arrayBufferToBase64(buffer) {
  const bytes = new Uint8Array(buffer)
  let binary = ''
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i])
  }
  return btoa(binary)
}

async function getKey() {
  const keyB64 = import.meta.env.VITE_ENCRYPTION_KEY || ''
  if (!keyB64) {
    throw new Error('ENCRYPTION_KEY not set')
  }
  const keyData = base64ToArrayBuffer(keyB64)
  return crypto.subtle.importKey('raw', keyData, ALGORITHM, false, ['encrypt', 'decrypt'])
}

export async function encrypt(data) {
  const key = await getKey()
  const encoder = new TextEncoder()
  const plaintext = encoder.encode(JSON.stringify(data))
  const nonce = crypto.getRandomValues(new Uint8Array(12))
  const ciphertext = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: nonce },
    key,
    plaintext
  )
  const combined = new Uint8Array(nonce.length + ciphertext.byteLength)
  combined.set(nonce, 0)
  combined.set(new Uint8Array(ciphertext), nonce.length)
  return arrayBufferToBase64(combined.buffer)
}

export async function decrypt(token) {
  const key = await getKey()
  const decoder = new TextDecoder()
  const raw = base64ToArrayBuffer(token)
  const nonce = new Uint8Array(raw.slice(0, 12))
  const ciphertext = raw.slice(12)
  const plaintext = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: nonce },
    key,
    ciphertext
  )
  return JSON.parse(decoder.decode(plaintext))
}
